plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevControllerExtensions.kt

```diff
@@ -99,38 +118,104 @@ suspend fun FeatureDevController.onCodeGeneration(session: Session, message: Str
             messenger.sendAnswer(
                 tabId = tabId,
                 messageType = FeatureDevMessageType.Answer,
-                message = if (remainingIterations == 0) {
+                message =
+                if (remainingIterations == 0) {
                     message("amazonqFeatureDev.code_generation.iteration_zero")
                 } else {
                     message(
                         "amazonqFeatureDev.code_generation.iteration_counts",
                         remainingIterations,
-                        totalIterations
+                        totalIterations,
                     )
-                }
+                },
             )
         }
 
         messenger.sendSystemPrompt(tabId = tabId, followUp = getFollowUpOptions(session.sessionState.phase))
 
         messenger.sendUpdatePlaceholder(tabId = tabId, newPlaceholder = message("amazonqFeatureDev.placeholder.after_code_generation"))
     } finally {
-        messenger.sendAsyncEventProgress(tabId = tabId, inProgress = false) // Finish processing the event
-        messenger.sendChatInputEnabledMessage(tabId = tabId, enabled = false) // Lock chat input until a follow-up is clicked.
-
+        if (session.sessionState.token
+                ?.token()
+                ?.isCancellationRequested == true
+        ) {
+            session.sessionState.token = GradleConnector.newCancellationTokenSource()
+        } else {
+            messenger.sendAsyncEventProgress(tabId = tabId, inProgress = false) // Finish processing the event
+            messenger.sendChatInputEnabledMessage(tabId = tabId, enabled = false) // Lock chat input until a follow-up is clicked.
+        }
         if (toolWindow != null && !toolWindow.isVisible) {
             notifyInfo(
                 title = message("amazonqFeatureDev.code_generation.notification_title"),
                 content = message("amazonqFeatureDev.code_generation.notification_message"),
                 project = getProject(),
-                notificationActions = listOf(openChatNotificationAction())
+                notificationActions = listOf(openChatNotificationAction()),
             )
         }
     }
 }
 
-private fun FeatureDevController.openChatNotificationAction() = NotificationAction.createSimple(
-    message("amazonqFeatureDev.code_generation.notification_open_link")
+private suspend fun FeatureDevController.disposeToken(
+    state: SessionState,
+    messenger: MessagePublisher,
+    tabId: String,
+    remainingIterations: Number?,
+    totalIterations: Number?,
```

Comment:
why wouldnt these always be `Int`? why do we allow this to be nullable?

Reply:
Updated, thanks